docs/custom-filters.adoc

```diff
@@ -45,23 +229,220 @@ Instances are specific to the channel between a single client and a single broke
+optionally declare a configuration class that Kroxylicious will populate (using Jackson) when loading your custom Filter.
+The module must package a `META-INF/services/io.kroxylicious.proxy.filter.FilterContributor` file containing the
+classname of the contributor into the JAR file.
 
-== Packaging filters
+A convenience is provided to simplify building FilterContributors, the {api-javadoc}/io/kroxylicious/proxy/service/BaseContributor.html[BaseContributor]. See the {github-sample-contributor}[SampleContributor] for
+an example implementation (and corresponding {github}/blob/main/kroxylicious-sample/src/main/resources/META-INF/services/io.kroxylicious.proxy.filter.FilterContributor[service metadata]).
 
-Filters are packaged as standard `.jar` files.
+For simple integration with the BaseContributor a Custom Protocol Filter may have:
 
-// TODO
+1. No Constructor
+2. A no-arguments Constructor
+3. A Constructor where the only parameter is a class that extends {api-javadoc}/io/kroxylicious/proxy/config/BaseConfig.html[BaseConfig].
 
-== Sample filters
+If you use style 3, you will supply a custom configuration class. This class must be deserializable using https://github.com/FasterXML/jackson[FasterXML/Jackson]
+we use this to load filter configuration from the Kroxylicious configuration file into your custom class.
 
-A collection of sample filters is available within the Kroxylicious repository for you to download, try out, and customise.
-You can find them https://github.com/kroxylicious/kroxylicious/tree/main/kroxylicious-sample[here] for a hands-on introduction to creating your own custom filters.
\ No newline at end of file
+For example in the kroxylicious-samples we have the {github}/blob/main/kroxylicious-sample/src/main/java/io/kroxylicious/sample/config/SampleFilterConfig.java[SampleFilterConfig] class.
+This is used in the {github-sample-contributor}[SampleContributor] when building up the mappings and then again in the
+constructors of the Filters (see {github}/blob/main/kroxylicious-sample/src/main/java/io/kroxylicious/sample/SampleFetchResponseFilter.java[SampleFetchResponseFilter]).
+
+Then, when we configure a filter in Kroxylicious configuration like:
+
+[source,yaml]
+----
+filters:
+- type: SampleFetchResponse
+  config:
+    findValue: a
+    replacementValue: b
+----
+Kroxylicious will deserialize the `config` object into a `SampleFilterConfig` and use it to construct a
+`SampleFetchResponseFilter` passing the `SampleFilterConfig` instance as a constructor argument.
+
+Note that we have the <<Composite Filters, CompositeFilter>> interface available if you wish to use a single configuration
+block in YAML to install multiple Filters into the Filter chain.
+
+== Packaging filters
+
+Filters are packaged as standard `.jar` files. A typical Custom Filter jar contains:
```

Comment:
Should we have some mention of class paths? Shading etc here for third-party dependencies?

(Happy for that to be a future PR too)

Reply:
Yeah, can leave packaging for another PR, it's a bit chunky.